velox/functions/sparksql/aggregates/FirstLastAggregate.cpp

```diff
@@ -0,0 +1,353 @@
+  }
+};
+
+template <bool numeric, bool ignoreNull, typename T>
+class LastAggregate : public FirstLastAggregateBase<numeric, T> {
+ public:
+  explicit LastAggregate(TypePtr resultType)
+      : FirstLastAggregateBase<numeric, T>(resultType) {}
+
+  void addRawInput(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /* mayPushdown */) override {
+    DecodedVector decoded(*args[0], rows);
+
+    if (decoded.isConstantMapping()) {
+      if (!decoded.isNullAt(0)) {
+        rows.applyToSelected(
+            [&](vector_size_t i) { this->updateValue(i, groups[i], decoded); });
+      }
+    } else if (decoded.mayHaveNulls()) {
+      rows.applyToSelected([&](vector_size_t i) {
+        if (!decoded.isNullAt(i)) {
+          this->updateValue(i, groups[i], decoded);
+        } else if (!ignoreNull) {
+          exec::Aggregate::setNull(groups[i]);
+        }
+      });
+    } else {
+      rows.applyToSelected(
+          [&](vector_size_t i) { this->updateValue(i, groups[i], decoded); });
+    }
+  }
+
+  void addIntermediateResults(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool mayPushdown) override {
+    addRawInput(groups, rows, args, mayPushdown);
+  }
+
+  void addSingleGroupRawInput(
+      char* group,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /* mayPushdown */) override {
+    DecodedVector decoded(*args[0], rows);
+
+    auto i = rows.end() - 1;
+    while (i >= rows.begin()) {
+      if (!rows.isValid(i)) {
+        --i;
+        continue;
+      }
+      if (!decoded.isNullAt(i)) {
+        this->updateValue(i, group, decoded);
+        return;
+      }
+      if (!ignoreNull) {
+        exec::Aggregate::setNull(group);
+        return;
+      }
+      --i;
+    }
+  }
+
+  void addSingleGroupIntermediateResults(
+      char* group,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool mayPushdown) override {
+    addSingleGroupRawInput(group, rows, args, mayPushdown);
+  }
+};
+
+} // namespace
+
+template <template <bool B1, bool B2, typename T> class TClass, bool ignoreNull>
+bool registerFirstLast(const std::string& name) {
+  std::vector<std::shared_ptr<exec::AggregateFunctionSignature>> signatures;
+  for (const auto& inputType :
+       {"tinyint",
+        "smallint",
+        "integer",
+        "bigint",
+        "real",
+        "double",
+        "varchar"}) {
+    signatures.push_back(exec::AggregateFunctionSignatureBuilder()
+                             .argumentType(inputType)
+                             .intermediateType(inputType)
+                             .returnType(inputType)
+                             .build());
+  }
+  signatures.push_back(exec::AggregateFunctionSignatureBuilder()
+                           .typeVariable("T")
```

Comment:
This signature implies that we support arrays of *all* types.

Reply:
What I'm trying to say is that currently the signatures are define in such a way that first(array(row(...))) is supported, but first(row(...)) is not. This doesn't quite makes sense.

Reply:
Similarly, first(array(boolean)) is supported, but first(boolean) is not.